File: 1122-relative-sort-array/1122-relative-sort-array.js
/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number[]}
 */
var relativeSortArray = function(arr1, arr2) {
    let HM = {};
    
    for(let index = 0; index < arr1.length; index++) {
        if(arr1[index] in HM) {
            HM[arr1[index]]++;
        } else {
            HM[arr1[index]] = 1;
        }
    }
    
    let result = [];
    let resultIndex = 0;
    
    for(let index = 0; index < arr2.length; index++) {
        for(let count = 0; count < HM[arr2[index]]; count++) {
            result[resultIndex++] = arr2[index];
        }
        HM[arr2[index]] = 0;
    }
    
    arr1.sort((a,b) => a - b);
    
    for(let index = 0; index < arr1.length; index++) {
        for(let count = 0; count < HM[arr1[index]]; count++) {
            result[resultIndex++] = arr1[index];
        }
        HM[arr1[index]] = 0;
    }
    
    return result;
};

File: 1122-relative-sort-array/1122-relative-sort-array.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        HashMap<Integer, Integer> HM = new HashMap<Integer, Integer>();
        
        for(int i = 0; i < arr1.length; i++) {
            if(HM.containsKey(arr1[i])) {
                HM.put(arr1[i], HM.get(arr1[i]) + 1);
            } else {
                HM.put(arr1[i], 1);
            }
        }
        
        int[] result = new int[arr1.length];
        int index = 0;
        
        for(int i = 0; i < arr2.length; i++) {
            for(int count = 0; count < HM.get(arr2[i]); count++) {
                result[index++] = arr2[i];
            }
            
            HM.put(arr2[i], 0);
        }
        
        Arrays.sort(arr1);
        
        for(int i = 0; i < arr1.length; i++) {
            if(HM.get(arr1[i]) > 0) {
                for(int count = 0; count < HM.get(arr1[i]); count++) {
                    result[index++] = arr1[i];
                }
                HM.put(arr1[i], 0);
            } 
        }
        
        return result;
    }
}